File: Linked List/doubly_circular/deletion/list.java
// Doubly Circular Linked List - Deletion (Java)
// Delete head and first occurrence by value.

class Node {
    int data;
    Node prev;
    Node next;
    Node(int value) { this.data = value; this.prev = null; this.next = null; }
}

public class list {
    static Node deleteHead(Node head) {
        if (head == null) return null;
        if (head.next == head) return null;
        Node tail = head.prev;
        Node newHead = head.next;
        tail.next = newHead; newHead.prev = tail;
        return newHead;
    }

    static Node deleteByValue(Node head, int target) {
        if (head == null) return null;
        Node cur = head;
        do {
            if (cur.data == target) {
                if (cur == head) return deleteHead(head);
                cur.prev.next = cur.next;
                cur.next.prev = cur.prev;
                return head;
            }
            cur = cur.next;
        } while (cur != head);
        return head;
    }

    static void printForward(Node head) {
        if (head == null) { System.out.println(); return; }
        Node cur = head; do { System.out.print(cur.data + " "); cur = cur.next; } while (cur != head); System.out.println();
    }

    public static void main(String[] args) {
        Node n1 = new Node(10), n2 = new Node(20), n3 = new Node(30);
        n1.next = n2; n2.prev = n1; n2.next = n3; n3.prev = n2; n3.next = n1; n1.prev = n3;
        Node head = n1;
        System.out.print("Initial: "); printForward(head);
        head = deleteHead(head); System.out.print("After deleteHead: "); printForward(head);
        head = deleteByValue(head, 30); System.out.print("After deleteByValue(30): "); printForward(head);
    }
}

File: Linked List/doubly_circular/deletion/list.py
# Doubly Circular Linked List - Deletion (Python)
# Delete head and first occurrence by value.

class Node:
    def __init__(self, value: int) -> None:
        self.data = value
        self.prev: "Node | None" = None
        self.next: "Node | None" = None


def delete_head(head: "Node | None") -> "Node | None":
    if head is None:
        return None
    if head.next is head:
        return None
    tail = head.prev
    new_head = head.next
    tail.next = new_head
    new_head.prev = tail
    return new_head


def delete_by_value(head: "Node | None", target: int) -> "Node | None":
    if head is None:
        return None
    cur = head
    while True:
        if cur.data == target:
            if cur is head:
                return delete_head(head)
            cur.prev.next = cur.next
            cur.next.prev = cur.prev
            return head
        cur = cur.next
        if cur is head:
            break
    return head


def print_forward(head: "Node | None") -> None:
    if head is None:
        print()
        return
    cur = head
    while True:
        print(cur.data, end=" ")
        cur = cur.next
        if cur is head:
            break
    print()


if __name__ == "__main__":
    n1 = Node(10); n2 = Node(20); n3 = Node(30)
    n1.next = n2; n2.prev = n1
    n2.next = n3; n3.prev = n2
    n3.next = n1; n1.prev = n3
    head = n1
    print("Initial:", end=" "); print_forward(head)
    head = delete_head(head); print("After delete_head:", end=" "); print_forward(head)
    head = delete_by_value(head, 30); print("After delete_by_value(30):", end=" "); print_forward(head)


